helix-lsp/src/lib.rs

```diff
@@ -648,48 +670,46 @@ impl Registry {
         doc_path: Option<&std::path::PathBuf>,
         root_dirs: &[PathBuf],
         enable_snippets: bool,
-    ) -> Result<Option<Arc<Client>>> {
-        let config = match &language_config.language_server {
-            Some(config) => config,
-            None => return Ok(None),
-        };
-
-        let scope = language_config.scope.clone();
-
-        match self.inner.entry(scope) {
-            Entry::Vacant(_) => Ok(None),
-            Entry::Occupied(mut entry) => {
-                // initialize a new client
-                let id = self.counter.fetch_add(1, Ordering::Relaxed);
-
-                let NewClientResult(client, incoming) = start_client(
-                    id,
-                    language_config,
-                    config,
-                    doc_path,
-                    root_dirs,
-                    enable_snippets,
-                )?;
-                self.incoming.push(UnboundedReceiverStream::new(incoming));
-
-                let old_clients = entry.insert(vec![(id, client.clone())]);
-
-                for (_, old_client) in old_clients {
-                    tokio::spawn(async move {
-                        let _ = old_client.force_shutdown().await;
-                    });
+    ) -> Result<Vec<Arc<Client>>> {
+        language_config
+            .language_servers
+            .iter()
+            .filter_map(|LanguageServerFeatures { name, .. }| {
+                if self.inner.contains_key(name) {
+                    let client = match self.start_client(
+                        name.clone(),
+                        language_config,
+                        doc_path,
+                        root_dirs,
+                        enable_snippets,
+                    ) {
+                        Ok(client) => client,
+                        error => return Some(error),
+                    };
+                    let old_clients = self
+                        .inner
+                        .insert(name.clone(), vec![client.clone()])
+                        .unwrap();
+
+                    // TODO what if there are multiple instances for different workspaces?
+                    // I think the language servers will be stopped without being restarted, which is not intended
```

Comment:
Ok, well yeah now I see it. As long as `refresh_language_servers` is called afterwards, it's fine indeed, but it may make sense to put a doc-comment on top of the function (`restart`) that explains that, as it's a public method of the crate, and whenever it's used, it might make sense to remind the user, that this has some side effects that are not immediately visible by the method syntax.

Reply:
I've added a doc-comment in https://github.com/helix-editor/helix/pull/2507/commits/a76b9b5999604675975033dc802bce89082933e1, I think this is resolved then.